Fixes regression added while fixing Bug #72740

scara opened the pull request:
HI @0mars,
https://github.com/php/pecl-search_engine-solr/commit/b1b44e0a22bcce8f625707248d0b4d3630935359#diff-ab0834017b5c60ae382501bce9a32bbc added a breaking regression when using Query Fields qith no boost value due to a "NUL byte issue".
It has been there since 2.5 years ago: most of - i.e. not all - the package maintainers in several distro were still using 2.4.0 with just few patches for 7.2 compat.
Test coverage was unfortunately incomplete to highlight the `#0;` issue in the `qf` parameter.

_2.5.0_ is finally out and using it under the condition above breaks any expected response; the improved test now shows the issue:
```diff
# make test TESTS="--show-all tests/bug_72740.phpt"
[cut]
========DIFF========
016+     [qf] => score#0; manufacturedate_dt#0;
016-     [qf] => score^ manufacturedate_dt^
========DONE========
```

Long story short, I discovered this issue since months in a project called Moodle within the context of its CI Docker Toolbox: you can read more in https://github.com/moodlehq/moodle-php-apache/issues/16 and https://github.com/moodlehq/moodle-php-apache/issues/19 but limited spare time and the availability of workarounds kept me far from proposing kind of fix, waiting for the release of 2.5.0.

On 7/7 I found some spare time to explore the code and to propose a quick fix to highlight where the problem was (https://github.com/scara/pecl-search_engine-solr/commit/c2c94b459ceab41df93cfa8a635f0578aef73a83) and posted it into https://bugs.php.net/bug.php?id=75631: @glensc was kind enough to do a quick peer review (really appreciated!).
I've decided to keep the initial quick fix in this PR since it looks like the way the code is actually managing the _NULL byte check_, **even** if @glensc is **super right** in suggesting a better refactoring which IMHO should be done at any level in the code to remove the pattern of checking for the value to be `NULL` and then via `strlen() > 0` since it has been initialized with an empty string i.e. `\0`.

For the record, https://bugs.php.net/bug.php?id=72740 is still open even if it has been already "addressed".

HTH,
Matteo

glensc commented on src/php5/solr_functions_params.c:
```diff
@@ -714,7 +714,9 @@ PHP_SOLR_API void solr_arg_list_param_value_fetch(solr_param_t *solr_param, solr
+		    }
 		} else {
 		    solr_string_appendc(&tmp_buffer, separator);
 		}
```
I think you need to combine these if-conditions, or add another else. first one is preferred, because second one just adds duplicate code.

1. combined
```c
		if (current_ptr->contents.arg_list.delimiter_override && strlen(current_ptr->contents.arg_list.delimiter_override) > 0) {
		    solr_string_appendc(&tmp_buffer, *current_ptr->contents.arg_list.delimiter_override);
		} else {
		    solr_string_appendc(&tmp_buffer, separator);
		}
```

2. else for second if

```c
		if (current_ptr->contents.arg_list.delimiter_override) {
		    if (strlen(current_ptr->contents.arg_list.delimiter_override) > 0) {
		        solr_string_appendc(&tmp_buffer, *current_ptr->contents.arg_list.delimiter_override);
		    } else {
		        solr_string_appendc(&tmp_buffer, separator);
		    }
		} else {
		    solr_string_appendc(&tmp_buffer, separator);
		}

```

scara replied:
TNX @glensc!
Would you prefer a new cleanup - 3rd - commit or just amending the current one?

TIA,
Matteo

glensc replied:
While this does not apply fully, as changes are small and buffered in my memory, but:

As the review process is in order; you should add commits; it's rather annoying to review if the commits have been squashed in middle of review; i.e have to review again the whole codebase, can't know which is old and which is new change; you can tidy up the branch later when you find it's ready.

I personally use [git fixup](https://github.com/keis/git-fixup) tool aid `git rebase --autosquash`

scara replied:
Hi @glensc,
pushed a 3rd commit, TNX for the suggestion/fix.

+1 for squashing all the commits into a single one, before merging it into the mainstream.

HTH,
Matteo

glensc commented on src/php5/solr_functions_params.c:
```diff
@@ -713,10 +713,8 @@ PHP_SOLR_API void solr_arg_list_param_value_fetch(solr_param_t *solr_param, solr
-		    if (strlen(current_ptr->contents.arg_list.delimiter_override) > 0) {
-		        solr_string_appendc(&tmp_buffer, *current_ptr->contents.arg_list.delimiter_override);
-		    }
+		if (current_ptr->contents.arg_list.delimiter_override && strlen(current_ptr->contents.arg_list.delimiter_override) > 0) {
```
the length calculation is not actually needed, it's only needed to check the value is not empty. considering it is [asciiz](https://en.wikipedia.org/wiki/Null-terminated_string) string, the index=0 should not be `\0`:

```c
if (current_ptr->contents.arg_list.delimiter_override && *current_ptr->contents.arg_list.delimiter_override) {
```

or:

```c
if (current_ptr->contents.arg_list.delimiter_override && current_ptr->contents.arg_list.delimiter_override[0]) {
```

whichever you find more readable.

glensc replied:
seems rest of the code uses first variant.

scara replied:
Hi @glensc,
OK, will do this late evening.

I'll add a new test scenario too, the one testing [`addQueryField()`](https://php.net/manual/en/solrdismaxquery.addqueryfield.php) w/ a `boost` value, the one failing using my "quick fix" instead of your "tidy up suggestion" which was more than that i.e. a fix in my broken logic 😊.

TNX for your patience, appreciated 😉 .

scara commented:
Hi @glensc,
if you're fine w/ the current code I'd squash the commits into two: the first for the test changes and the latter for the code changes.

TIA,
Matteo

glensc commented:
@scara LGTM; altho as expressed elsewhere, it would be cleaner if `current_ptr->contents.arg_list.delimiter_override` are not assigned empty values (value with length=0) but rather `NULL`. but this can be separate PR as there may be other values that need to be simplified like this.

scara commented:
TNX @glensc 

I agree w/ you:
1. the "bad thing" is using an empty string instead of `NULL`. Guessing some historical reason or an initial different code design
1. it would be nice to refactor the code, granted a good coverage of each refactored piece of code. Too big for this PR: besides, this extension is actually broken whenever the PHP code will use `addQueryField()` and this PR should be IMHO merged ASAP and 2.5.1 shipped

Will do the planned squashing late my evening.
Then, I'll open a specific issue describing how it would be useful such a refactoring and how it could be error-prone to stick with the empty string init pattern.

HTH,
Matteo

scara commented:
Hi @glensc,
I've squashed and forced the push: PR is now clean TNX to your assistance 👍.

I've added a ref to this PR even in the original bug, https://bugs.php.net/bug.php?id=72740#1562881713.

HTH,
Matteo

scara commented:
Hi @glensc,
is there something I can do to promote this PR?
Should I contact some package maintainers at least to share the issue in _2.5.0_ and what workaround can be adopted to mitigate it?

TIA,
Matteo

glensc commented:
@scara I'm nobody in pecl/php/pecl-solr scope, sorry ;) you need to reach maintainers somehow.

I can only encourage maintainers (with this comment) that this PR looks ok for me and should be merged into 2.5.1 patch release

0mars commented:
Thanks a lot, will follow up next weekend

scara commented:
Hi @Jan-E,
I'm reading about your interest here from #13: is there anything I can do to help you in merging this PR?

scara commented:
Still wondering why after 7 months the PR is still at this stage, without a peer review from the maintainers.

BTW TNX to @glensc for the time spent on helping me, really appreciated!

Matteo

stronk7 commented:
Hi, is there any scheduled planned for this, say 2.6.0 or 2.5.1 or anything?

Having to build solr everywhere with the patch applied, because the upstream version simply doesn't work ok, or having to force boost values as a workaround aren't really the best options.

That said by an ignorant that knows nothing about the internals, just the experience is suboptimal, I'm afraid.

Many thanks for all the work with this awesome extension, much appreciated!

Ciao :-)

php-pulls commented:
**Comment on behalf of omars at php.net:**

merged
